src/test/scala/com/wavesplatform/UtxPoolSpecification.scala

```diff
@@ -0,0 +1,118 @@
+    val genesisSettings = TestHelpers.genesisSettings(Map(senderAccount -> senderBalance))
+    val (_, _, state, bcu) =
+      BlockStorageImpl(BlockchainSettings(None, None, None, 'T', 5, FunctionalitySettings.TESTNET, genesisSettings)).get
+
+    bcu.processBlock(Block.genesis(genesisSettings).right.get)
+
+    state
+  }
+
+  private def transfer(sender: PrivateKeyAccount, maxAmount: Long, time: Time) = (for {
+    amount <- chooseNum(1, (maxAmount * 0.9).toLong)
+    recipient <- accountGen
+    fee <- chooseNum(1, (maxAmount * 0.1).toLong)
+  } yield TransferTransaction.create(None, sender, recipient, amount, time.getTimestamp(), None, fee, Array.empty[Byte]).right.get)
+    .label("transferTransaction")
+
+  private def expectBroadcast(tx: Transaction): Unit =
+    (group.writeAndFlush(_: Any, _: ChannelMatcher)).expects(RawBytes(25, tx.bytes), ChannelMatchers.all()).once()
+
+  private val stateGen = for {
+    sender <- accountGen.label("sender")
+    senderBalance <- positiveLongGen.label("senderBalance")
+  } yield (sender, senderBalance, mkState(sender, senderBalance))
+
+  private def utxTest(utxSettings: UtxSettings = UtxSettings(20, 5.seconds), txCount: Int = 10)
+                     (f: (Seq[TransferTransaction], UtxPool, TestTime) => Unit): Unit = forAll(
+    stateGen,
+    chooseNum(2, txCount).label("txCount")) { case ((sender, senderBalance, state), count) =>
+      val time = new TestTime().reset()
+
+      forAll(listOfN(count, transfer(sender, senderBalance / 2, time))) { txs =>
+        val utx = new UtxPool(group, time, state, calculator, FunctionalitySettings.TESTNET, utxSettings)
+        f(txs, utx, time)
+      }
+    }
+
+  private val zzgen: Gen[(UtxPool, TestTime, Seq[Transaction], FiniteDuration, Seq[Transaction])] =
+    for {
+      (sender, senderBalance, state) <- stateGen
+      ts = System.currentTimeMillis()
+      count1 <- chooseNum(5, 10)
+      tx1 <- listOfN(count1, transfer(sender, senderBalance / 2, new TestTime(ts)))
+      offset <- chooseNum(5000L, 10000L)
+      tx2 <- listOfN(count1, transfer(sender, senderBalance / 2, new TestTime(ts + offset + 1000)))
+    } yield {
+      val time = new TestTime(System.currentTimeMillis())
+      val utx = new UtxPool(group, time, state, calculator, FunctionalitySettings.TESTNET, UtxSettings(10, offset.millis))
+      (utx, time, tx1, (offset + 1000).millis, tx2)
+    }
+
+  "UTX Pool" - {
+    "does not add new transactions when full" in utxTest(UtxSettings(1, 5.seconds)) { (txs, utx, _) =>
+      expectBroadcast(txs.head)
+      utx.putIfNew(txs.head) shouldBe 'right
+      all(txs.tail.map(t => utx.putIfNew(t))) shouldBe 'left
+    }
+
+    "does not add the same transaction twice" in utxTest() { (txs, utx, _) =>
+      expectBroadcast(txs.head)
+      utx.putIfNew(txs.head) shouldBe 'right
+      utx.putIfNew(txs.head) shouldEqual Left(GenericError(s"Transaction ${txs.head.id} already in the pool"))
```

Comment:
Consider `should produde("already in the pool") as less strict and easier to assert